File: platforms/apple/Sources/BotaAppleSDK/Bluetooth/EncryptedUploadV2TransferControl.swift
import Foundation

actor EncryptedUploadV2TransferControl {
    typealias Subscribe = @Sendable (String) async throws -> AsyncThrowingStream<Data, Error>
    typealias Write = @Sendable (String, Data) async throws -> Void
    typealias Unsubscribe = @Sendable (String) async throws -> Void

    private static let startMessageType: UInt8 = 0x20
    private static let resumeRequestMessageType: UInt8 = 0x22
    private static let cleanupAbortReason: UInt16 = 0x00FF
    private static let defaultReplyTimeoutNanoseconds: UInt64 = 10_000_000_000
    private static let defaultCleanupTimeoutNanoseconds: UInt64 = 1_000_000_000

    private let mapper: CoreModelMapper
    private let subscribe: Subscribe
    private let write: Write
    private let unsubscribe: Unsubscribe
    private var exchangeActive = false
    private var activeTransfer: ActiveEncryptedUploadV2Transfer?
    private var cleanupUncertain = false

    init(
        mapper: CoreModelMapper,
        subscribe: @escaping Subscribe,
        write: @escaping Write,
        unsubscribe: @escaping Unsubscribe
    ) {
        self.mapper = mapper
        self.subscribe = subscribe
        self.write = write
        self.unsubscribe = unsubscribe
    }

    init(bluetooth: CoreBluetoothHost, mapper: CoreModelMapper) {
        self.init(
            mapper: mapper,
            subscribe: { peripheralID in
                try await bluetooth.subscribe(
                    peripheralID: peripheralID,
                    serviceUUID: BotaBluetoothUUIDs.storageService,
                    characteristicUUID: BotaBluetoothUUIDs.recordingTransferV2
                )
            },
            write: { peripheralID, data in
                try await bluetooth.write(
                    peripheralID: peripheralID,
                    serviceUUID: BotaBluetoothUUIDs.storageService,
                    characteristicUUID: BotaBluetoothUUIDs.transferControlV2,
                    data: data
                )
            },
            unsubscribe: { peripheralID in
                try await bluetooth.unsubscribe(
                    peripheralID: peripheralID,
                    serviceUUID: BotaBluetoothUUIDs.storageService,
                    characteristicUUID: BotaBluetoothUUIDs.recordingTransferV2
                )
            }
        )
    }

    func start(
        peripheralID: String,
        request: EncryptedUploadV2StartRequestValue,
        replyTimeoutNanoseconds: UInt64 = defaultReplyTimeoutNanoseconds,
        cleanupTimeoutNanoseconds: UInt64 = defaultCleanupTimeoutNanoseconds
    ) async throws -> EncryptedUploadV2StartAcknowledgementValue {
        let frame = try mapper.createEncryptedUploadV2Start(
            transportSessionID: request.transportSessionID,
            uploadSessionID: request.uploadSessionID,
            recordingUUID: request.recordingUUID,
            recordingGeneration: request.recordingGeneration,
            authorizationSHA256: request.authorizationSHA256,
            checkpointRevision: request.checkpointRevision,
            nextCiphertextOffset: request.nextCiphertextOffset,
            prefixSHA256: request.prefixSHA256,
            windowPackets: request.windowPackets,
            dataPayloadBytes: request.dataPayloadBytes
        )
        let terminal: MatchedTransferControlReply<EncryptedUploadV2StartAcknowledgementValue> = try await exchange(
            peripheralID: peripheralID,
            frame: frame,
            transportSessionID: request.transportSessionID,
            failedMessageType: Self.startMessageType,
            replyTimeoutNanoseconds: replyTimeoutNanoseconds,
            cleanupTimeoutNanoseconds: cleanupTimeoutNanoseconds,
            retainSubscription: { _ in true }
        ) { value in
            guard case let .startAccepted(acknowledgement) = value else { return nil }
            guard acknowledgement.uploadSessionID == request.uploadSessionID,
                  Self.sameUUID(acknowledgement.recordingUUID, request.recordingUUID),
                  acknowledgement.recordingGeneration == request.recordingGeneration,
                  acknowledgement.ciphertextLength == request.expectedCiphertextLength,
                  acknowledgement.ciphertextSHA256 == request.expectedCiphertextSHA256,
                  acknowledgement.windowPackets == request.windowPackets,
                  acknowledgement.dataPayloadBytes == request.dataPayloadBytes,
                  acknowledgement.checkpointIntervalBlocks == request.expectedCheckpointIntervalBlocks,
                  acknowledgement.checkpointRevision == request.checkpointRevision,
                  acknowledgement.nextCiphertextOffset == request.nextCiphertextOffset,
                  acknowledgement.prefixSHA256 == request.prefixSHA256
            else {
                throw Self.error(
                    code: .identityMismatch,
                    detail: "START_ACK does not match the requested recording, ciphertext, or checkpoint"
                )
            }
            return acknowledgement
        }
        switch terminal {
        case let .value(acknowledgement):
            return acknowledgement
        case let .deviceError(error):
            throw Self.deviceRejection(error)
        }
    }

    func resume(
        peripheralID: String,
        request: EncryptedUploadV2ResumeRequestValue,
        replyTimeoutNanoseconds: UInt64 = defaultReplyTimeoutNanoseconds,
        cleanupTimeoutNanoseconds: UInt64 = defaultCleanupTimeoutNanoseconds
    ) async throws -> EncryptedUploadV2ResumeDecision {
        let frame = try mapper.createEncryptedUploadV2ResumeRequest(
            transportSessionID: request.transportSessionID,
            uploadSessionID: request.uploadSessionID,
            recordingUUID: request.recordingUUID,
            recordingGeneration: request.recordingGeneration,
            checkpointRevision: request.checkpointRevision,
            nextCiphertextOffset: request.nextCiphertextOffset,
            prefixSHA256: request.prefixSHA256,
            windowPackets: request.windowPackets,
            dataPayloadBytes: request.dataPayloadBytes
        )
        let terminal: MatchedTransferControlReply<EncryptedUploadV2ResumeDecision> = try await exchange(
            peripheralID: peripheralID,
            frame: frame,
            transportSessionID: request.transportSessionID,
            failedMessageType: Self.resumeRequestMessageType,
            replyTimeoutNanoseconds: replyTimeoutNanoseconds,
            cleanupTimeoutNanoseconds: cleanupTimeoutNanoseconds,
            retainSubscription: { decision in
                if case .accepted = decision { return true }
                return false
            }
        ) { value in
            switch value {
            case let .resumeAccepted(accepted):
                guard accepted.uploadSessionID == request.uploadSessionID,
                      Self.sameUUID(accepted.recordingUUID, request.recordingUUID),
                      accepted.recordingGeneration == request.recordingGeneration,
                      accepted.checkpointRevision == request.checkpointRevision,
                      accepted.nextCiphertextOffset == request.nextCiphertextOffset,
                      accepted.prefixSHA256 == request.prefixSHA256,
                      accepted.windowPackets == request.windowPackets,
                      accepted.dataPayloadBytes == request.dataPayloadBytes
                else {
                    throw Self.error(
                        code: .identityMismatch,
                        detail: "RESUME_ACCEPT does not match the requested recording or checkpoint"
                    )
                }
                return .accepted(accepted)
            case let .resumeRejected(rejected):
                guard rejected.reason != 0 else {
                    throw Self.error(
                        code: .unexpectedEvent,
                        detail: "RESUME_REJECT cannot carry a success result"
                    )
                }
                return .rejected(rejected)
            case .startAccepted, .error:
                return nil
            }
        }
        switch terminal {
        case let .value(decision):
            return decision
        case let .deviceError(error):
            throw Self.deviceRejection(error)
        }
    }

    func resetAfterConfirmedDisconnect() {
        cleanupUncertain = false
        activeTransfer = nil
    }

    func claimNotificationStream(
        transportSessionID: UInt64
    ) throws -> AsyncThrowingStream<Data, Error> {
        guard var transfer = activeTransfer,
              transfer.transportSessionID == transportSessionID
        else {
            throw Self.error(
                code: .identityMismatch,
                detail: "no active encrypted transfer matches the requested transport session"
            )
        }
        guard !transfer.notificationStreamClaimed else {
            throw Self.error(
                code: .operationInProgress,
                detail: "the active encrypted transfer notification stream is already claimed"
            )
        }
        transfer.notificationStreamClaimed = true
        activeTransfer = transfer
        return transfer.notifications
    }

    func writeActiveTransferFrame(
        transportSessionID: UInt64,
        frame: Data
    ) async throws {
        guard let transfer = activeTransfer,
              transfer.transportSessionID == transportSessionID,
              transfer.notificationStreamClaimed
        else {
            throw Self.error(
                code: .identityMismatch,
                detail: "no claimed encrypted transfer matches the requested transport session"
            )
        }
        guard frame.count >= 68,
              frame.first == 0x21,
              Self.readUInt64(frame, at: 4) == transportSessionID
        else {
            throw Self.error(
                code: .invalidInput,
                detail: "active encrypted transfer frame is malformed or belongs to another session"
            )
        }
        try await write(transfer.peripheralID, frame)
    }

    func abortActiveTransfer(
        transportSessionID: UInt64,
        reason: UInt16,
        cleanupTimeoutNanoseconds: UInt64 = defaultCleanupTimeoutNanoseconds
    ) async throws {
        guard !exchangeActive else {
            throw Self.error(
                code: .operationInProgress,
                detail: "an encrypted transfer-control exchange is still active"
            )
        }
        guard let transfer = activeTransfer,
              transfer.transportSessionID == transportSessionID
        else {
            throw Self.error(
                code: .identityMismatch,
                detail: "no active encrypted transfer matches the requested transport session"
            )
        }
        exchangeActive = true
        defer { exchangeActive = false }

        let abort = try mapper.createEncryptedUploadV2Abort(
            transportSessionID: transportSessionID,
            reason: reason
        )
        let cleaned = await cleanup(
            peripheralID: transfer.peripheralID,
            abort: abort,
            timeoutNanoseconds: cleanupTimeoutNanoseconds
        )
        activeTransfer = nil
        guard cleaned else {
            throw Self.error(
                code: .uploadOwnershipUnknown,
                detail: "active encrypted transfer cleanup is uncertain"
            )
        }
    }

    private func exchange<Value: Sendable>(
        peripheralID: String,
        frame: Data,
        transportSessionID: UInt64,
        failedMessageType: UInt8,
        replyTimeoutNanoseconds: UInt64,
        cleanupTimeoutNanoseconds: UInt64,
        retainSubscription: @escaping @Sendable (Value) -> Bool,
        match: @escaping @Sendable (EncryptedUploadV2TransferControlValue) throws -> Value?
    ) async throws -> MatchedTransferControlReply<Value> {
        guard !exchangeActive, activeTransfer == nil else {
            throw Self.error(
                code: .operationInProgress,
                detail: "another encrypted transfer-control owner is active"
            )
        }
        guard !cleanupUncertain else {
            throw Self.error(
                code: .uploadOwnershipUnknown,
                detail: "transfer-control cleanup is uncertain; reconnect before retrying"
            )
        }
        exchangeActive = true
        defer { exchangeActive = false }

        let abort = try mapper.createEncryptedUploadV2Abort(
            transportSessionID: transportSessionID,
            reason: Self.cleanupAbortReason
        )
        let notifications = try await subscribe(peripheralID)
        var sent = false
        var terminalReleasesDevice = false
        var cleanupAttempted = false

        do {
            let reply = try await withThrowingTaskGroup(
                of: MatchedTransferControlReply<Value>.self
            ) { group in
                group.addTask {
                    for try await data in notifications {
                        let value = try self.mapper.decodeEncryptedUploadV2TransferControl(data)
                        guard value.transportSessionID == transportSessionID else {
                            throw Self.error(
                                code: .identityMismatch,
                                detail: "encrypted transfer reply belongs to another transport session"
                            )
                        }
                        if case let .error(error) = value {
                            guard error.failedMessageType == failedMessageType,
                                  error.result != 0
                            else {
                                throw Self.error(
                                    code: .unexpectedEvent,
                                    detail: "encrypted transfer ERROR does not match the active request"
                                )
                            }
                            return .deviceError(error)
                        }
                        guard let matched = try match(value) else {
                            throw Self.error(
                                code: .unexpectedEvent,
                                detail: "encrypted transfer reply type does not match the active request"
                            )
                        }
                        return .value(matched)
                    }
                    try Task.checkCancellation()
                    throw Self.error(
                        code: .unexpectedEvent,
                        detail: "transfer-control notification stream ended without a matching reply"
                    )
                }
                do {
                    try Task.checkCancellation()
                    sent = true
                    try await write(peripheralID, frame)
                    group.addTask {
                        try await Task.sleep(nanoseconds: replyTimeoutNanoseconds)
                        throw Self.error(
                            code: .timeout,
                            retryable: true,
                            detail: "timed out waiting for the matching transfer-control reply"
                        )
                    }
                    guard let reply = try await group.next() else {
                        throw Self.error(
                            code: .unexpectedEvent,
                            detail: "transfer-control reply race ended unexpectedly"
                        )
                    }
                    group.cancelAll()
                    return reply
                } catch {
                    group.cancelAll()
                    throw error
                }
            }
            let retain = switch reply {
            case let .value(value): retainSubscription(value)
            case .deviceError: false
            }
            if retain {
                try Task.checkCancellation()
                activeTransfer = ActiveEncryptedUploadV2Transfer(
                    peripheralID: peripheralID,
                    transportSessionID: transportSessionID,
                    notifications: notifications
                )
                return reply
            }
            terminalReleasesDevice = true
            try Task.checkCancellation()
            cleanupAttempted = true
            guard await cleanup(
                peripheralID: peripheralID,
                abort: nil,
                timeoutNanoseconds: cleanupTimeoutNanoseconds
            ) else {
                throw Self.error(
                    code: .uploadOwnershipUnknown,
                    detail: "transfer-control subscription cleanup is uncertain"
                )
            }
            return reply
        } catch {
            if !cleanupAttempted {
                _ = await cleanup(
                    peripheralID: peripheralID,
                    abort: sent && !terminalReleasesDevice ? abort : nil,
                    timeoutNanoseconds: cleanupTimeoutNanoseconds
                )
            }
            throw error
        }
    }

    private func cleanup(
        peripheralID: String,
        abort: Data?,
        timeoutNanoseconds: UInt64
    ) async -> Bool {
        let write = self.write
        let unsubscribe = self.unsubscribe
        let completed = await Self.boundedCleanup(timeoutNanoseconds: timeoutNanoseconds) {
            var firstError: Error?
            if let abort {
                do {
                    try await write(peripheralID, abort)
                } catch {
                    firstError = error
                }
            }
            do {
                try await unsubscribe(peripheralID)
            } catch {
                if firstError == nil { firstError = error }
            }
            if let firstError { throw firstError }
        }
        if !completed { cleanupUncertain = true }
        return completed
    }

    private static func boundedCleanup(
        timeoutNanoseconds: UInt64,
        operation: @escaping @Sendable () async throws -> Void
    ) async -> Bool {
        await withCheckedContinuation { continuation in
            let race = TransferControlCleanupRace(continuation)
            Task.detached {
                do {
                    try await operation()
                    race.resolve(true)
                } catch {
                    race.resolve(false)
                }
            }
            Task.detached {
                try? await Task.sleep(nanoseconds: timeoutNanoseconds)
                race.resolve(false)
            }
        }
    }

    private static func sameUUID(_ lhs: String, _ rhs: String) -> Bool {
        guard let lhs = UUID(uuidString: lhs), let rhs = UUID(uuidString: rhs) else { return false }
        return lhs == rhs
    }

    private static func readUInt64(_ data: Data, at offset: Int) -> UInt64 {
        data[offset..<(offset + 8)].enumerated().reduce(0) { value, pair in
            value | (UInt64(pair.element) << UInt64(pair.offset * 8))
        }
    }

    private static func deviceRejection(
        _ error: EncryptedUploadV2TransferErrorValue
    ) -> EncryptedUploadV2TransferControlRejection {
        EncryptedUploadV2TransferControlRejection(
            sdkError: Self.error(
                code: .protocolRejected,
                protocolStatus: error.result,
                detail: "device rejected the encrypted transfer-control request"
            ),
            deviceError: error
        )
    }

    private static func error(
        code: BotaSDKErrorCode,
        retryable: Bool = false,
        protocolStatus: UInt16? = nil,
        detail: String
    ) -> BotaSDKError {
        BotaSDKError(
            code: code,
            operation: .transferRecording,
            retryable: retryable,
            protocolStatus: protocolStatus,
            detail: detail
        )
    }
}

private struct ActiveEncryptedUploadV2Transfer: Sendable {
    let peripheralID: String
    let transportSessionID: UInt64
    let notifications: AsyncThrowingStream<Data, Error>
    var notificationStreamClaimed = false
}

private enum MatchedTransferControlReply<Value: Sendable>: Sendable {
    case value(Value)
    case deviceError(EncryptedUploadV2TransferErrorValue)
}

private extension EncryptedUploadV2TransferControlValue {
    var transportSessionID: UInt64 {
        switch self {
        case let .startAccepted(value): value.transportSessionID
        case let .resumeAccepted(value): value.transportSessionID
        case let .resumeRejected(value): value.transportSessionID
        case let .error(value): value.transportSessionID
        }
    }
}

private final class TransferControlCleanupRace: @unchecked Sendable {
    private let lock = NSLock()
    private var continuation: CheckedContinuation<Bool, Never>?

    init(_ continuation: CheckedContinuation<Bool, Never>) {
        self.continuation = continuation
    }

    func resolve(_ value: Bool) {
        lock.lock()
        let continuation = self.continuation
        self.continuation = nil
        lock.unlock()
        continuation?.resume(returning: value)
    }
}

File: platforms/android/sdk/src/main/kotlin/dev/bota/sdk/internal/bluetooth/EncryptedUploadV2TransferControl.kt
package dev.bota.sdk.internal.bluetooth

import dev.bota.sdk.internal.core.CoreModelMapper
import dev.bota.sdk.internal.core.EncryptedUploadV2ResumeRequest
import dev.bota.sdk.internal.core.EncryptedUploadV2StartRequest
import dev.bota.sdk.internal.core.EncryptedUploadV2TransferControlValue
import dev.bota.sdk.internal.core.EncryptedUploadV2TransferPayload
import dev.bota.sdk.internal.host.EncryptedUploadV2HostException
import java.security.MessageDigest
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.CoroutineStart
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.consumeAsFlow
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withTimeout
import kotlinx.coroutines.TimeoutCancellationException

internal sealed interface EncryptedUploadV2OpenResult {
    data class Opened(val notifications: Flow<EncryptedUploadV2TransferPayload>) : EncryptedUploadV2OpenResult
    data object ResumeRejected : EncryptedUploadV2OpenResult
}

internal class EncryptedUploadV2TransferControl(
    private val driver: BluetoothDriver,
    private val mapper: CoreModelMapper,
    private val controlTimeoutMilliseconds: Long = ControlTimeoutMilliseconds,
) : AutoCloseable {
    private data class Session(
        val peripheralId: String,
        val raw: Channel<ByteArray>,
        val job: Job,
    )

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val mutex = Mutex()
    private val sessions = mutableMapOf<ULong, Session>()

    suspend fun open(
        peripheralId: String,
        request: EncryptedUploadV2StartRequest,
        checkpoint: EncryptedUploadV2CheckpointValue?,
    ): EncryptedUploadV2OpenResult {
        val ready = CompletableDeferred<Unit>()
        val raw = Channel<ByteArray>(capacity = 2_048)
        val job = scope.launch(start = CoroutineStart.LAZY) {
            try {
                val notifications = driver.subscribe(
                    peripheralId,
                    BotaBluetoothUUIDs.StorageService,
                    BotaBluetoothUUIDs.RecordingTransferV2,
                )
                ready.complete(Unit)
                notifications.collect { raw.send(it.value) }
                raw.close()
            } catch (error: Throwable) {
                ready.completeExceptionally(error)
                raw.close(error)
            }
        }
        val session = Session(peripheralId, raw, job)
        try {
            mutex.withLock {
                require(request.transportSessionId != 0uL && sessions.isEmpty()) {
                    "encrypted transfer session is already active"
                }
                sessions[request.transportSessionId] = session
            }
            job.start()
            ready.await()
            val frame = if (checkpoint == null) {
                mapper.createEncryptedUploadV2Start(
                    request.transportSessionId, request.uploadSessionId, request.recordingUuid,
                    request.recordingGeneration, request.authorizationSha256, request.checkpointRevision,
                    request.nextCiphertextOffset, request.prefixSha256, request.windowPackets, request.dataPayloadBytes,
                )
            } else {
                mapper.createEncryptedUploadV2ResumeRequest(
                    EncryptedUploadV2ResumeRequest(
                        request.transportSessionId, request.uploadSessionId, request.recordingUuid,
                        request.recordingGeneration, checkpoint.revision, checkpoint.nextCiphertextOffset,
                        checkpoint.prefixSha256, request.windowPackets, request.dataPayloadBytes,
                    ),
                )
            }
            write(peripheralId, frame)
            val response = try {
                withTimeout(controlTimeoutMilliseconds) { raw.receive() }
            } catch (_: TimeoutCancellationException) {
                throw EncryptedUploadV2HostException(
                    15u, true, message = "timed out waiting for the encrypted transfer acknowledgement",
                )
            }
            when (val control = mapper.decodeEncryptedUploadV2TransferControl(response)) {
                is EncryptedUploadV2TransferControlValue.StartAccepted -> {
                    expected(checkpoint == null, "START received an unexpected reply type")
                    validateStart(request, control.value)
                }
                is EncryptedUploadV2TransferControlValue.ResumeAccepted -> {
                    expected(checkpoint != null, "RESUME received an unexpected reply type")
                    validateResume(request, checkpoint, control.value)
                }
                is EncryptedUploadV2TransferControlValue.ResumeRejected -> {
                    identity(
                        control.value.transportSessionId == request.transportSessionId,
                        "RESUME_REJECT belongs to another transport session",
                    )
                    expected(checkpoint != null && control.value.reason != 0.toUShort(), "RESUME_REJECT is invalid")
                    releaseSession(request.transportSessionId, session)
                    return EncryptedUploadV2OpenResult.ResumeRejected
                }
                is EncryptedUploadV2TransferControlValue.Error -> {
                    identity(
                        control.value.transportSessionId == request.transportSessionId,
                        "transfer ERROR belongs to another transport session",
                    )
                    val expectedMessage = if (checkpoint == null) 0x20.toUByte() else 0x22.toUByte()
                    expected(
                        control.value.result != 0.toUShort() && control.value.failedMessageType == expectedMessage,
                        "transfer ERROR does not match the active request",
                    )
                    throw EncryptedUploadV2HostException(
                        17u, false, control.value.result,
                        "device rejected the encrypted transfer",
                    )
                }
            }
            return EncryptedUploadV2OpenResult.Opened(
                raw.consumeAsFlow().map(mapper::decodeEncryptedUploadV2TransferPayload),
            )
        } catch (error: Throwable) {
            releaseSession(request.transportSessionId, session)
            throw error
        }
    }

    suspend fun writeActiveFrame(transportSessionId: ULong, frame: ByteArray) = mutex.withLock {
        val session = sessions[transportSessionId] ?: error("encrypted transfer session is not active")
        write(session.peripheralId, frame)
    }

    suspend fun abort(transportSessionId: ULong, reason: UShort = 0x00ffu.toUShort()) {
        val session = mutex.withLock { sessions[transportSessionId] } ?: return
        runCatching { write(session.peripheralId, mapper.createEncryptedUploadV2Abort(transportSessionId, reason)) }
        releaseSession(transportSessionId, session)
    }

    suspend fun release(transportSessionId: ULong) {
        val session = mutex.withLock { sessions[transportSessionId] } ?: return
        releaseSession(transportSessionId, session)
    }

    override fun close() {
        scope.coroutineContext[Job]?.cancel()
        sessions.clear()
    }

    private suspend fun releaseSession(id: ULong, session: Session) {
        mutex.withLock { sessions.remove(id, session) }
        session.job.cancel()
        session.raw.close()
        runCatching {
            driver.unsubscribe(
                session.peripheralId,
                BotaBluetoothUUIDs.StorageService,
                BotaBluetoothUUIDs.RecordingTransferV2,
            )
        }
    }

    private suspend fun write(peripheralId: String, frame: ByteArray) {
        require(frame.size <= minOf(driver.maximumWriteLength(peripheralId), 512)) {
            "Rust-generated transfer frame exceeds negotiated write length"
        }
        driver.write(
            peripheralId,
            BotaBluetoothUUIDs.StorageService,
            BotaBluetoothUUIDs.TransferControlV2,
            frame,
            withResponse = true,
        )
    }

    private fun validateStart(request: EncryptedUploadV2StartRequest, value: dev.bota.sdk.internal.core.EncryptedUploadV2StartAcknowledgement) {
        identity(
            value.transportSessionId == request.transportSessionId && value.uploadSessionId == request.uploadSessionId &&
                value.recordingUuid == request.recordingUuid && value.recordingGeneration == request.recordingGeneration &&
                value.ciphertextLength == request.expectedCiphertextLength &&
                MessageDigest.isEqual(value.ciphertextSha256, request.expectedCiphertextSha256) &&
                value.windowPackets == request.windowPackets && value.dataPayloadBytes == request.dataPayloadBytes &&
                value.checkpointIntervalBlocks == request.expectedCheckpointIntervalBlocks &&
                value.checkpointRevision == request.checkpointRevision &&
                value.nextCiphertextOffset == request.nextCiphertextOffset &&
                MessageDigest.isEqual(value.prefixSha256, request.prefixSha256),
            "START acknowledgement does not match the selected encrypted transfer",
        )
    }

    private fun validateResume(
        request: EncryptedUploadV2StartRequest,
        checkpoint: EncryptedUploadV2CheckpointValue?,
        value: dev.bota.sdk.internal.core.EncryptedUploadV2ResumeValue,
    ) {
        identity(checkpoint != null && value.transportSessionId == request.transportSessionId &&
            value.uploadSessionId == request.uploadSessionId && value.recordingUuid == request.recordingUuid &&
            value.recordingGeneration == request.recordingGeneration && value.checkpointRevision == checkpoint.revision &&
            value.nextCiphertextOffset == checkpoint.nextCiphertextOffset &&
            MessageDigest.isEqual(value.prefixSha256, checkpoint.prefixSha256) &&
            value.windowPackets == request.windowPackets && value.dataPayloadBytes == request.dataPayloadBytes,
            "RESUME acknowledgement does not match the durable checkpoint",
        )
    }

    private fun identity(condition: Boolean, detail: String) {
        if (!condition) throw EncryptedUploadV2HostException(11u, false, message = detail)
    }

    private fun expected(condition: Boolean, detail: String) {
        if (!condition) throw EncryptedUploadV2HostException(9u, false, message = detail)
    }

    private companion object {
        const val ControlTimeoutMilliseconds = 10_000L
    }
}
